src/ClassDiscovery.php

```diff
@@ -2,126 +2,117 @@
     {
-        if (null === self::$puliFactory) {
-            if (!defined('PULI_FACTORY_CLASS')) {
-                throw new \RuntimeException('Puli Factory is not available');
-            }
-
-            $puliFactoryClass = PULI_FACTORY_CLASS;
+        // Look in the cache
+        if (null !== $class = self::getFromCache($type)) {
+            return $class;
+        }
 
-            if (!class_exists($puliFactoryClass)) {
-                throw new \RuntimeException('Puli Factory class does not exist');
+        $exceptions = [];
+        foreach (self::$strategies as $strategy) {
+            try {
+                $candidates = call_user_func($strategy.'::getCandidates', $type);
+            } catch (StrategyUnavailableException $e) {
+                $exceptions[] = $e;
+                continue;
             }
 
-            self::$puliFactory = new $puliFactoryClass();
-        }
+            foreach ($candidates as $candidate) {
+                if (isset($candidate['condition'])) {
+                    if (!self::evaluateCondition($candidate['condition'])) {
+                        continue;
+                    }
+                }
 
-        return self::$puliFactory;
-    }
+                // save the result for later use
+                self::storeInCache($type, $candidate);
 
-    /**
-     * Sets the Puli factory.
-     *
-     * @param object $puliFactory
-     */
-    public static function setPuliFactory($puliFactory)
-    {
-        if (!is_callable([$puliFactory, 'createRepository']) || !is_callable([$puliFactory, 'createDiscovery'])) {
-            throw new \InvalidArgumentException('The Puli Factory must expose a repository and a discovery');
+                return $candidate['class'];
+            }
         }
 
-        self::$puliFactory = $puliFactory;
-        self::$puliDiscovery = null;
+        throw new DiscoveryFailedException('Could not find resource using any discovery strategy', $exceptions);
     }
 
     /**
-     * Resets the factory.
-     */
-    public static function resetPuliFactory()
-    {
-        self::$puliFactory = null;
-        self::$puliDiscovery = null;
-    }
-
-    /**
-     * Returns the Puli discovery layer.
+     * Get a value from cache.
+     *
+     * @param string $type
      *
-     * @return Discovery
+     * @return string|null
      */
-    public static function getPuliDiscovery()
+    private static function getFromCache($type)
     {
-        if (!isset(self::$puliDiscovery)) {
-            $factory = self::getPuliFactory();
-            $repository = $factory->createRepository();
+        if (!isset(self::$cache[$type])) {
+            return;
+        }
 
-            self::$puliDiscovery = $factory->createDiscovery($repository);
+        $candidate = self::$cache[$type];
+        if (!self::evaluateCondition($candidate['condition'])) {
+            return;
         }
 
-        return self::$puliDiscovery;
+        return $candidate['class'];
```

Comment:
Do we need to cache the condition then? Maybe it's enough to cache only the class?